replica/compaction_group.hh

```diff
@@ -260,9 +260,17 @@ public:
     // FIXME: Cannot return nullptr, signature can be changed to return storage_group&.
     storage_group* storage_group_for_id(const schema_ptr&, size_t i) const;
 
-    // Caller must keep the current effective_replication_map_ptr valid
-    // until the storage_group_manager finishes update_effective_replication_map
-    virtual future<> update_effective_replication_map(const locator::effective_replication_map& erm) = 0;
+    class erm_updater {
+    public:
+        virtual ~erm_updater() {}
+        // Caller must keep the current effective_replication_map_ptr valid
+        // until the storage_group_manager finishes update_effective_replication_map
+        virtual void update(const locator::effective_replication_map& erm) = 0;
+
+        // Waits for background work initiated by non-yielding update.
+        virtual future<> close() = 0;
+    };
+    virtual std::unique_ptr<erm_updater> make_erm_updater() = 0;
```

Comment:
Doesn't the original have just two implementations? Is it error-prone?

Reply:
Could be made simpler by passing update_effective_replication_map() a function.

Reply:
> Doesn't the original have just two implementations? Is it error-prone?

_sg_manager->update_effective_replication_map() cannot yield before updating its internal state, so the interface must be explicit about that (i.e. not return a future), but we also need to wait for work, so they were decoupled from each other. I am really scared about the fragility of the current interface.

Reply:
> Could be made simpler by passing update_effective_replication_map() a function.

to give storage_group_manager::update_effective_replication_map() control when new_erm is applied? that make sense.